Upgrade to llvm/clang 3.7

layus opened the pull request:
With this set of fixes, I am able to build mozart2 against llvm/clang 3.7.
This in turns allows to make builds for the new tcl/tk around.

Of course the documentation needs to be updated, but the previous pull request is still pending...

sjrd commented on vm/generator/main/generator.cc:
```diff
@@ -77,7 +77,7 @@ void processDeclContext(const std::string outputDir, const DeclContext* ds,
 
 int main(int argc, char* argv[]) {
-  llvm::IntrusiveRefCntPtr<DiagnosticsEngine> Diags;
+  llvm::IntrusiveRefCntPtr<DiagnosticsEngine> Diags = CompilerInstance::createDiagnostics(new DiagnosticOptions);
```
That's a long line. Maybe wrap it after `=`.

sjrd commented on vm/generator/main/builtins.cc:
```diff
@@ -223,9 +223,9 @@ void handleBuiltinModule(const std::string& outputDir, const ClassDecl* CD,
-    llvm::raw_fd_ostream to((outputDir+name+"-builtin.json").c_str(), err);
-    assert(err == "");
+    std::error_code err;
+    llvm::raw_fd_ostream to((outputDir+name+"-builtin.json").c_str(), err, llvm::sys::fs::F_None);
```
Maybe wrap after the new `,`?

sjrd commented on vm/generator/main/generator.cc:
```diff
@@ -105,9 +105,11 @@ int main(int argc, char* argv[]) {
-                                           false, 0, 0, true);
+  clang::CompilerInstance CI;
+  std::unique_ptr<ASTUnit> unit = ASTUnit::LoadFromASTFile(astFile,
+                                           PCHContainerOperations().getRawReader(),
```
Wrapped arguments should align with the the first argument. However, here it becomes very long, so maybe wrap directly after `(`, in which case indent 2 (or 4? I don't remember, best check somewhere else we did this) spaces wrt the `s` of `std::`.

sjrd commented:
I guess this makes sense.

sjrd commented:
Should this be merged with #231?

layus commented:
Updated according to the comments.

#231 is about upgrading to 3.5, this one is about upgrading to 3.7.
The differences are
- between 3.5 and 3.7, LoadFromASTFile also changed.
- #231 also updates the docs and the build process to 3.5, this one is more lazy.

Maybe we should merge this with #231, but IMHO it makes little sense to maintain a build process fetching llvm sources.

Just my two cents, but we really should ship partially built sources.
Providing an archive with the result of `make gensources genboostsources library_stage_0` would remove the need for LLVM/Clang. It would shorten and simplify building and packaging mozart on any platform.

sjrd commented:
LGTM. Feel free to merge this PR and/or #231, whatever makes sense at this point.

layus commented:
Ah, but I do not have commit access. (Not sure I want it though... that's a great responsibility :wink:)

sjrd commented:
Well you're the only UCL member here, so you definitely should have them ^^

allentiak commented:
So... Who has commit access to merge this pull request?

allentiak commented:
Thanks! Should also #231 be closed?

sjmackenzie commented:
@allentiak closed.
